tests/kafkatest/tests/copycat_distributed_test.py

```diff
@@ -76,19 +79,103 @@ def test_file_source_and_sink(self, converter="org.apache.kafka.copycat.json.Jso
+                self.logger.info("Cleanly bouncing Copycat on " + str(node.account))
+                self.cc.stop_node(node)
+                with node.account.monitor_log(self.cc.LOG_FILE) as monitor:
+                    self.cc.start_node(node)
+                    monitor.wait_until("Starting connectors and tasks using config offset", timeout_sec=90,
+                                       err_msg="Copycat worker didn't successfully join group and start work")
+                self.logger.info("Bounced Copycat on %s and rejoined in %f seconds", node.account, time.time() - started)
+
+        self.source.stop()
+        self.sink.stop()
+        self.cc.stop()
+
+        # Validate at least once delivery of everything that was reported as written since we should have flushed and
+        # cleanly exited. Currently this only tests at least once delivery because the sink task may not have consumed
+        # all the messages generated by the source task. This needs to be done per-task since seqnos are not unique across
+        # tasks.
+        src_msgs = self.source.messages()
+        sink_msgs = self.sink.messages()
+        success = True
+        errors = []
+        for task in range(num_tasks):
+            src_seqnos = [msg['seqno'] for msg in src_msgs if msg['task'] == task]
+            # Every seqno up to the largest one we ever saw should appear. Each seqno should only appear once because clean
+            # bouncing should commit on rebalance.
+            src_seqno_max = max(src_seqnos)
+            src_seqno_counts = Counter(src_seqnos)
+            missing_src_seqnos = sorted(set(range(src_seqno_max)).difference(set(src_seqnos)))
+            duplicate_src_seqnos = sorted([seqno for seqno,count in src_seqno_counts.iteritems() if count > 1])
+
+            if missing_src_seqnos:
+                self.logger.error("Missing source sequence numbers for task " + str(task))
+                errors.append("Found missing source sequence numbers for task %d: %s" % (task, missing_src_seqnos))
+                success = False
+            if duplicate_src_seqnos:
+                self.logger.error("Duplicate source sequence numbers for task " + str(task))
+                errors.append("Found duplicate source sequence numbers for task %d: %s" % (task, duplicate_src_seqnos))
+                success = False
+
+            sink_seqnos = [msg['seqno'] for msg in sink_msgs if msg['task'] == task and 'flushed' in msg]
+            # Every seqno up to the largest one we ever saw should appear. Each seqno should only appear once because
+            # clean bouncing should commit on rebalance.
+            sink_seqno_max = max(sink_seqnos)
+            sink_seqno_counts = Counter(sink_seqnos)
+            missing_sink_seqnos = sorted(set(range(sink_seqno_max)).difference(set(sink_seqnos)))
+            duplicate_sink_seqnos = sorted([seqno for seqno,count in sink_seqno_counts.iteritems() if count > 1])
+
+            if missing_sink_seqnos:
```

Comment:
@guozhangwang @benstopford We can't guarantee they are exactly the same. The problem is deciding when it is safe to stop the sink connector. Currently we just kill it quickly after killing the source, so we can't be sure all the data generated by the source has been processed.

We could do something like tail the output of all the nodes running copycat workers, waiting until the last message produced is flushed by a sync. I wasn't convinced this was worth a) the more complicated test code and b) the fact that you then have to stream in realtime all the output of the workers.